File: apps/mobile/lib/shared/attachment/entities/share_reference.dart
library;

/// SHARE CONTRACT ALIGNMENT V1
///
/// Share = Reference Distribution Layer
/// - Share does NOT copy objects
/// - Share does NOT mutate source
/// - Share creates a reference with cached preview data
///
/// Reference Types:
/// - content: Universal social content
/// - for_sale: Fixed-price sale surfaces
/// - auction: Active auctions
/// - profile: User profiles
///
/// SEMANTIC RULES:
/// - targetType + targetId is the CANONICAL reference
/// - Preview data (title, imageUrl) is CACHE for UI only
/// - Business logic MUST resolve through backend using canonical IDs
/// - Preview data can be stale - never use for business decisions

import 'package:equatable/equatable.dart';
import 'package:labuda/shared/object/object_preview.dart' as obj;

// ============================================================================
// Enums
// ============================================================================

/// Types of entities that can be shared
enum ShareTargetType {
  content('content', 'content', '/content', 'contents', 'Content'),
  forSale('for_sale', 'for_sale', '/for-sale', 'for-sales', 'Produk Dijual'),
  auction('auction', 'auction', '/auction', 'auctions', 'Lelang'),
  profile('profile', 'profile', '/user', 'users', 'Profil');

  const ShareTargetType(
    this.wireValue,
    this.objectType,
    this.navigationPath,
    this.apiPath,
    this.displayName,
  );

  final String wireValue;
  final String objectType;
  final String navigationPath;
  final String apiPath;
  final String displayName;

  static ShareTargetType? fromString(String value) {
    for (final type in ShareTargetType.values) {
      if (type.wireValue == value) {
        return type;
      }
    }
    return null;
  }
}

// ============================================================================
// Value Objects
// ============================================================================

// ============================================================================
// Main Entity
// ============================================================================

/// ShareReference represents a reference to shared content.
/// It contains the canonical reference and cached preview data.
///
/// CONTRACT:
/// - targetType + targetId is the CANONICAL reference to the entity
/// - Use targetId for all backend queries and business logic
/// - Preview data is cached for UI display only and may be stale
///
/// **SEPARATION CLEANUP:** ShareReference is now independent from Attachment system.
/// It's a pure transport object for cross-domain references, not part of the legacy
/// Attachment wrapper hierarchy.
class ShareReference extends Equatable {
  /// The type of entity being referenced
  final ShareTargetType targetType;

  /// The canonical ID of the entity being referenced
  /// USE THIS for all backend queries
  final String targetId;

  /// Wire-level target type used when serializing for API/chat transport.
  /// Defaults to [targetType.wireValue], but chat content shares can override this
  /// to an older content wire value without changing the logical domain type.
  final String wireTargetType;

  /// Cached preview data for UI display
  /// DO NOT use for business decisions
  final obj.ObjectPreview preview;

  ShareReference({
    required this.targetType,
    required this.targetId,
    String? wireTargetType,
    required this.preview,
  }) : wireTargetType = wireTargetType ?? targetType.wireValue;

  // ============================================================================
  // Factory Constructors
  // ============================================================================

  /// Factory to create a share reference for content
  factory ShareReference.content({
    required String contentId,
    required String title,
    String? imageUrl,
    String? wireTargetType,
    bool isAvailable = true,
    bool isSold = false,
    bool isClosed = false,
    bool isDeleted = false,
  }) {
    return ShareReference(
      targetType: ShareTargetType.content,
      targetId: contentId,
      wireTargetType: wireTargetType,
      preview: obj.ObjectPreview(
        id: contentId,
        type: 'content',
        title: title,
        imageUrl: imageUrl,
        status: isAvailable ? 'available' : 'unavailable',
        isAvailable: isAvailable,
        isSold: isSold,
        isClosed: isClosed,
        isDeleted: isDeleted,
      ),
    );
  }

  /// Factory to create a share reference for a fixed-price sale
  factory ShareReference.forSale({
    required String forSaleId,
    required String title,
    String? imageUrl,
    String? wireTargetType,
    bool isAvailable = true,
    bool isSold = false,
    bool isClosed = false,
    bool isDeleted = false,
  }) {
    return ShareReference(
      targetType: ShareTargetType.forSale,
      targetId: forSaleId,
      wireTargetType: wireTargetType,
      preview: obj.ObjectPreview(
        id: forSaleId,
        type: ShareTargetType.forSale.objectType,
        title: title,
        imageUrl: imageUrl,
        status: isSold ? 'sold' : (isAvailable ? 'available' : 'unavailable'),
        isAvailable: isAvailable,
        isSold: isSold,
        isClosed: isClosed,
        isDeleted: isDeleted,
      ),
    );
  }

  /// Factory to create a share reference for auction
  factory ShareReference.auction({
    required String auctionId,
    required String title,
    String? imageUrl,
    String? wireTargetType,
    bool isAvailable = true,
    bool isSold = false,
    bool isClosed = false,
    bool isDeleted = false,
  }) {
    return ShareReference(
      targetType: ShareTargetType.auction,
      targetId: auctionId,
      wireTargetType: wireTargetType,
      preview: obj.ObjectPreview(
        id: auctionId,
        type: 'auction',
        title: title,
        imageUrl: imageUrl,
        status: isClosed ? 'ended' : (isAvailable ? 'active' : 'unavailable'),
        isAvailable: isAvailable,
        isSold: isSold,
        isClosed: isClosed,
        isDeleted: isDeleted,
      ),
    );
  }

  /// Factory to create a share reference for profile
  factory ShareReference.profile({
    required String profileId,
    required String name,
    String? avatarUrl,
    String? wireTargetType,
    bool isAvailable = true,
    bool isSold = false,
    bool isClosed = false,
    bool isDeleted = false,
  }) {
    return ShareReference(
      targetType: ShareTargetType.profile,
      targetId: profileId,
      wireTargetType: wireTargetType,
      preview: obj.ObjectPreview(
        id: profileId,
        type: 'profile',
        title: name,
        imageUrl: avatarUrl,
        status: isAvailable ? 'available' : 'unavailable',
        isAvailable: isAvailable,
        isSold: isSold,
        isClosed: isClosed,
        isDeleted: isDeleted,
      ),
    );
  }

  // ============================================================================
  // Business Logic
  // ============================================================================

  /// Check if this share reference has a valid target ID
  bool get isValid => targetId.isNotEmpty;

  /// Check if this share reference has an image in preview
  bool get hasImage => preview.imageUrl != null && preview.imageUrl!.isNotEmpty;

  /// Get the internal navigation path for this share reference.
  String get navigationPath => '${targetType.navigationPath}/$targetId';

  /// Get the object preview type used by the live preview resolvers.
  String get objectType => targetType.objectType;

  /// Get the API path for this share reference
  String get apiPath => '${targetType.apiPath}/$targetId';

  /// Wire target type that is safe to emit into chat payloads.
  ///
  /// Canonical content references emit `content` directly.
  String? get chatWireTargetType {
    if (targetType != ShareTargetType.content) {
      return wireTargetType;
    }

    return wireTargetType == 'content' ? wireTargetType : null;
  }

  /// Returns a ShareReference normalized for chat transport, or null when
  /// the reference is not valid for chat payloads.
  ShareReference? asChatReference() {
    final chatWireType = chatWireTargetType;
    if (chatWireType == null) {
      return null;
    }
    if (chatWireType == wireTargetType) {
      return this;
    }
    return copyWith(wireTargetType: chatWireType);
  }

  /// Get the display name for this share reference
  String get displayName => targetType.displayName;

  @override
  List<Object?> get props => [targetType, targetId, wireTargetType, preview];

  ShareReference copyWith({
    ShareTargetType? targetType,
    String? targetId,
    String? wireTargetType,
    obj.ObjectPreview? preview,
  }) {
    return ShareReference(
      targetType: targetType ?? this.targetType,
      targetId: targetId ?? this.targetId,
      wireTargetType: wireTargetType ?? this.wireTargetType,
      preview: preview ?? this.preview,
    );
  }

  /// Convert to JSON for API requests
  Map<String, dynamic> toJson() {
    return {
      'targetType': wireTargetType,
      'targetId': targetId,
      'preview': {
        'title': preview.title,
        'imageUrl': preview.imageUrl,
        'isAvailable': preview.isAvailable,
        'isSold': preview.isSold,
        'isClosed': preview.isClosed,
        'isDeleted': preview.isDeleted,
      },
    };
  }

  /// Create from JSON
  ///
  /// ZERO LEGACY MODE: Invalid targetType will fail - no fallback
  factory ShareReference.fromJson(Map<String, dynamic> json) {
    final wireTargetType = json['targetType'] as String;
    final targetType = _logicalTargetTypeFromWire(wireTargetType);
    final targetId = json['targetId'] as String;
    final previewJson = json['preview'] as Map<String, dynamic>?;
    final isAvailable = previewJson?['isAvailable'] as bool? ?? true;
    final isSold = previewJson?['isSold'] as bool? ?? false;
    final isClosed = previewJson?['isClosed'] as bool? ?? false;
    final isDeleted = previewJson?['isDeleted'] as bool? ?? false;

    // Determine status from flags
    String status;
    switch (targetType) {
      case ShareTargetType.forSale:
        status = isSold ? 'sold' : (isAvailable ? 'available' : 'unavailable');
        break;
      case ShareTargetType.auction:
        status = isClosed ? 'ended' : (isAvailable ? 'active' : 'unavailable');
        break;
      case ShareTargetType.content:
      case ShareTargetType.profile:
        status = isAvailable ? 'available' : 'unavailable';
        break;
    }

    return ShareReference(
      targetType: targetType,
      targetId: targetId,
      wireTargetType: wireTargetType,
      preview: obj.ObjectPreview(
        id: targetId,
        type: targetType.objectType,
        title: previewJson?['title'] ?? '',
        imageUrl: previewJson?['imageUrl'] as String?,
        status: status,
        isAvailable: isAvailable,
        isSold: isSold,
        isClosed: isClosed,
        isDeleted: isDeleted,
      ),
    );
  }
}

ShareTargetType _logicalTargetTypeFromWire(String wireTargetType) {
  final parsed = ShareTargetType.fromString(wireTargetType);
  if (parsed != null) {
    return parsed;
  }
  throw FormatException('Invalid share target type: $wireTargetType');
}

File: backend/internal/social/content/entity/share_reference.go
package entity

import (
	"fmt"
)

// SHARE CONTRACT ALIGNMENT V1:
//
// Share = Reference Distribution Layer
// - Share does NOT copy objects
// - Share does NOT mutate source
// - Share creates a reference with cached preview data
//
// Reference Types:
// - content: Social posts and requests
// - for_sale: Fixed-price sale surfaces
// - auction: Active auctions
// - profile: User profiles
//
// SEMANTIC RULES:
// - targetType + targetId is the CANONICAL reference
// - Preview data (title, imageUrl) is CACHE for UI only
// - Business logic MUST resolve through backend using canonical IDs
// - Preview data can be stale - never use for business decisions

// ShareTargetType defines the type of entity being shared.
type ShareTargetType string

const (
	ShareTargetTypeContent        ShareTargetType = "content"
	ShareTargetTypeForSale ShareTargetType = "for_sale"
	ShareTargetTypeAuction        ShareTargetType = "auction"
	ShareTargetTypeProfile        ShareTargetType = "profile"
)

// SharePreview contains cached preview data for display.
// This data can be stale - use only for UI, never for business logic.
//
// SNAPSHOT HONESTY: Live status fields indicate current state of shared target.
// UI should use these fields to show appropriate indicators (sold, closed, unavailable).
type SharePreview struct {
	Title    string `json:"title"`
	ImageURL string `json:"imageUrl,omitempty"`

	// SNAPSHOT HONESTY FIELDS: Live status at time of share creation
	// These enable honest UI rendering without additional API calls
	IsAvailable bool `json:"isAvailable"` // true if fixed-price sale/auction is available
	IsSold      bool `json:"isSold"`      // true if fixed-price sale is sold
	IsClosed    bool `json:"isClosed"`    // true if auction is closed
	IsDeleted   bool `json:"isDeleted"`   // true if target is deleted
}

// ShareReference represents a reference to shared content.
// It contains the canonical reference and cached preview data.
type ShareReference struct {
	TargetType ShareTargetType `json:"targetType"`
	TargetID   string          `json:"targetId"`
	Preview    SharePreview    `json:"preview"`
}

// NewShareReference creates a new share reference.
func NewShareReference(targetType ShareTargetType, targetID string, preview SharePreview) *ShareReference {
	return &ShareReference{
		TargetType: targetType,
		TargetID:   targetID,
		Preview:    preview,
	}
}

// NewShareReferenceFromContent creates a share reference from content.
//
// LIVE SHARE TRUTH V1: Accepts live status parameters for honest UI rendering.
// Callers MUST provide actual current status from content entity, not defaults.
func NewShareReferenceFromContent(
	contentID string,
	title string,
	imageURL string,
	isDeleted bool,
) *ShareReference {
	return &ShareReference{
		TargetType: ShareTargetTypeContent,
		TargetID:   contentID,
		Preview: SharePreview{
			Title:       title,
			ImageURL:    imageURL,
			IsAvailable: !isDeleted, // Content is available if not deleted
			IsSold:      false,      // Content doesn't have sold state
			IsClosed:    false,      // Content doesn't have closed state
			IsDeleted:   isDeleted,
		},
	}
}

// NewShareReferenceFromForSale creates a share reference from a fixed-price sale.
//
// LIVE SHARE TRUTH V1: Accepts live status parameters for honest UI rendering.
// Callers MUST provide actual current status from the fixed-price sale entity, not defaults.
//
// Expected source of truth: fixed-price sale availability and status fields.
func NewShareReferenceFromForSale(
	forSaleID string,
	title string,
	imageURL string,
	isAvailable bool,
	isSold bool,
	isDeleted bool,
) *ShareReference {
	return &ShareReference{
		TargetType: ShareTargetTypeForSale,
		TargetID:   forSaleID,
		Preview: SharePreview{
			Title:       title,
			ImageURL:    imageURL,
			IsAvailable: isAvailable,
			IsSold:      isSold,
			IsClosed:    false, // Fixed-price sales don't have closed state
			IsDeleted:   isDeleted,
		},
	}
}

// NewShareReferenceFromAuction creates a share reference from an auction.
//
// LIVE SHARE TRUTH V1: Accepts live status parameters for honest UI rendering.
// Callers MUST provide actual current status from auction entity, not defaults.
//
// Expected source of truth: Auction.Status field.
// - StatusActive -> IsAvailable: true, IsClosed: false
// - StatusEnded -> IsAvailable: false, IsClosed: true
// - StatusCancelled -> IsAvailable: false, IsClosed: true
func NewShareReferenceFromAuction(
	auctionID string,
	title string,
	imageURL string,
	isAvailable bool,
	isClosed bool,
	isDeleted bool,
) *ShareReference {
	return &ShareReference{
		TargetType: ShareTargetTypeAuction,
		TargetID:   auctionID,
		Preview: SharePreview{
			Title:       title,
			ImageURL:    imageURL,
			IsAvailable: isAvailable,
			IsSold:      false, // Auctions don't have sold state (they close)
			IsClosed:    isClosed,
			IsDeleted:   isDeleted,
		},
	}
}

// NewShareReferenceFromProfile creates a share reference from a profile.
//
// LIVE SHARE TRUTH V1: Accepts live status parameters for honest UI rendering.
// Callers MUST provide actual current status from profile/user entity.
func NewShareReferenceFromProfile(
	profileID string,
	name string,
	avatarURL string,
	isDeleted bool,
) *ShareReference {
	return &ShareReference{
		TargetType: ShareTargetTypeProfile,
		TargetID:   profileID,
		Preview: SharePreview{
			Title:       name,
			ImageURL:    avatarURL,
			IsAvailable: !isDeleted, // Profile is available if not deleted
			IsSold:      false,      // Profiles don't have sold state
			IsClosed:    false,      // Profiles don't have closed state
			IsDeleted:   isDeleted,
		},
	}
}

// IsValid returns true if the share reference has valid target ID.
func (s *ShareReference) IsValid() bool {
	return s.TargetID != ""
}

// DeepLink returns the deep link path for this share reference.
func (s *ShareReference) DeepLink() string {
	switch s.TargetType {
	case ShareTargetTypeContent:
		return fmt.Sprintf("/content/%s", s.TargetID)
	case ShareTargetTypeForSale:
		return fmt.Sprintf("/forSale/%s", s.TargetID)
	case ShareTargetTypeAuction:
		return fmt.Sprintf("/auction/%s", s.TargetID)
	case ShareTargetTypeProfile:
		return fmt.Sprintf("/profile/%s", s.TargetID)
	default:
		return ""
	}
}

// DisplayName returns the display name for the target type.
func (s ShareTargetType) DisplayName() string {
	switch s {
	case ShareTargetTypeContent:
		return "Post"
	case ShareTargetTypeForSale:
		return "Produk Dijual"
	case ShareTargetTypeAuction:
		return "Lelang"
	case ShareTargetTypeProfile:
		return "Profil"
	default:
		return string(s)
	}
}

// IsValid returns true if the target type is valid.
func (s ShareTargetType) IsValid() bool {
	switch s {
	case ShareTargetTypeContent, ShareTargetTypeForSale, ShareTargetTypeAuction,
		ShareTargetTypeProfile:
		return true
	default:
		return false
	}
}

// ParseShareTargetType parses a string into ShareTargetType.
func ParseShareTargetType(s string) (ShareTargetType, error) {
	t := ShareTargetType(s)
	if !t.IsValid() {
		return "", fmt.Errorf("invalid share target type: %s", s)
	}
	return t, nil
}
